src/sparseml/pytorch/sparsification/distillation/modifier_per_layer.py

```diff
@@ -0,0 +1,419 @@
+                f"{distillation_teacher}. "
+                "To disable set to 'disable' and for self attention set to 'self'"
+            )
+
+        self._cached_student_output.clear()
+        self._cached_teacher_output.clear()
+        self._student_output_shapes.clear()
+        self._teacher_output_shapes.clear()
+        self._student_handles.clear()
+        self._teacher_handles.clear()
+        self._projection.clear()
+
+        cached_student_layers: Dict[str, torch.nn.Module] = {}
+        if self.student_layer_names is None:
+            _find_layers_by_type(module, cached_student_layers)
+            self.student_layer_names = list(cached_student_layers.keys())
+        else:
+            _find_layers_by_name(
+                module, self.student_layer_names, cached_student_layers
+            )
+        _LOGGER.info("Distilling student layers: %s", self.student_layer_names)
+
+        cached_teacher_layers: Dict[str, torch.nn.Module] = {}
+        if self.teacher_layer_names is None:
+            _find_layers_by_type(self._teacher, cached_teacher_layers)
+            self.teacher_layer_names = list(cached_teacher_layers.keys())
+        else:
+            _find_layers_by_name(
+                self._teacher, self.teacher_layer_names, cached_teacher_layers
+            )
+        _LOGGER.info("Distilling teacher layers: %s", self.teacher_layer_names)
+
+        if len(self.teacher_layer_names) != len(self.student_layer_names):
+            raise ValueError(
+                "Found different numbers of teacher and student layers to distill. "
+                "Set teacher_layer_names and student_layer_names explicitly."
+            )
+
+        for layer_name in cached_student_layers:
+            self._student_handles.append(
+                cached_student_layers[layer_name].register_forward_hook(
+                    _create_cache_output_hook(
+                        layer_name,
+                        self._cached_student_output,
+                        self._student_output_shapes,
+                    )
+                )
+            )
+
+        for layer_name in cached_teacher_layers:
+            self._teacher_handles.append(
+                cached_teacher_layers[layer_name].register_forward_hook(
+                    _create_cache_output_hook(
+                        layer_name,
+                        self._cached_teacher_output,
+                        self._teacher_output_shapes,
+                    )
+                )
+            )
+
+    def update(
+        self,
+        module: Module,
+        optimizer: torch.optim.Optimizer,
+        epoch: float,
+        steps_per_epoch: int,
+    ):
+        super().update(module, optimizer, epoch, steps_per_epoch)
+        if epoch >= self.end_epoch:
+            for handle in self._student_handles:
+                handle.remove()
+            for handle in self._teacher_handles:
+                handle.remove()
```

Comment:
@bfineran is this what you meant by removing them after end epoch? will this work?